tutorials/roostats/StandardHypoTestDemo.py

```diff
@@ -0,0 +1,484 @@
+        if nuisParsInPdf.getSize() == 0:
+            warn(
+                "StandardHypoTestDemo",
+                "Prior nuisance does not depend on nuisance parameters. They will be smeared in their full range",
+            )
+
+        # HybridCalculator
+        hypoCalc.ForcePriorNuisanceAlt(nuisPdf)
+        hypoCalc.ForcePriorNuisanceNull(nuisPdf)
+
+    # hypoCalc->ForcePriorNuisanceAlt(*sbModel->GetPriorPdf());
+    # hypoCalc->ForcePriorNuisanceNull(*bModel->GetPriorPdf());
+
+    sampler = hypoCalc.GetTestStatSampler()
+
+    if sampler and (calcType == 0 or calcType == 1):
+
+        # look if pdf is number counting or extended
+        if sbModel.GetPdf().canBeExtended():
+            if useNC:
+                warn("StandardHypoTestDemo", "Pdf is extended: but number counting flag is set: ignore it ")
+    else:
+        # for not extended pdf
+        if not useNC:
+            nEvents = int(data.numEntries())
+            ROOT.Info(
+                "StandardHypoTestDemo",
+                "Pdf is not extended: number of events to generate taken  from observed data set is {nEvents}",
+            )
+            sampler.SetNEventsPerToy(nEvents)
+        else:
+            ROOT.Info("StandardHypoTestDemo", "using a number counting pdf")
+            sampler.SetNEventsPerToy(1)
+
+    if data.isWeighted() and not generateBinned:
+        msgfmt = "Data set is weighted, nentries = {} and sum of weights = {:8.1f} but toy ".format(
+            data.numEntries(), data.sumEntries()
+        )
+        msgfmt += "generation is unbinned - it would be faster to set generateBinned to True\n"
+
+        ROOT.Info("StandardHypoTestDemo", msgfmt)
+
+    if generateBinned:
+        sampler.SetGenerateBinned(generateBinned)
+
+    # set the test statistic
+    if testStatType == 0:
+        sampler.SetTestStatistic(slrts)
+    if testStatType == 1:
+        sampler.SetTestStatistic(ropl)
+    if testStatType == 2 or testStatType == 3:
+        sampler.SetTestStatistic(profll)
+
+    htr = hypoCalc.GetHypoTest()
+    htr.SetPValueIsRightTail(True)
+    htr.SetBackgroundAsAlt(False)
+    htr.Print()  # how to get meaningful CLs at this point?
+
+    del sampler
+    del slrts
+    del ropl
+    del profll
+    collect()  # Trigger the garbage collector gc.collector()
+
+    if calcType != 2:
+        c1 = ROOT.TCanvas("myc1", "myc1")
+        plot = ROOT.RooStats.HypoTestPlot(htr, 100)
+        plot.SetLogYaxis(True)
+        plot.Draw()
+        c1.Update()
+        c1.Draw()
+        c1.SaveAs("StandardHypoTestDemo.png")
+    else:
+        print("Asymptotic results ")
+
+    # look at expected significances
+    # found median of S+B distribution
+    if calcType != 2:
+
+        altDist = htr.GetAltDistribution()
+        htExp = ROOT.RooStats.HypoTestResult("Expected Result")
+        htExp.Append(htr)
+        # find quantiles in alt (S+B) distribution
+        p = np.array([i for i in range(5)], dtype=float)
```

Comment:
@PPaye, FYI, if you use NumPy arrays, then they can be directly passed to C++ functions that expect `double *`. Like this, you don't have to go via `ctypes`.

More on this in the cppyy documentation: https://cppyy.readthedocs.io/en/latest/lowlevel.html#lowlevelview